File: apps/api/schemas/deal.py
"""Deal request/response schemas."""

import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from apps.api.models.enums import (
    CommissionPayoutStatus,
    CommissionType,
    DealStage,
    DealType,
)


class DealCreateRequest(BaseModel):
    deal_type: DealType
    customer_id: uuid.UUID
    property_id: uuid.UUID
    listing_id: uuid.UUID | None = None
    primary_agent_id: uuid.UUID
    secondary_agent_id: uuid.UUID | None = None
    lead_id: uuid.UUID | None = None
    transaction_value: Decimal
    transaction_currency: str
    expected_close_at: date | None = None
    notes: str | None = None
    commission_type: CommissionType | None = None
    commission_value: Decimal | None = None
    commission_override_reason: str | None = None
    primary_commission_pct: int = Field(default=100, ge=0, le=100)
    secondary_commission_pct: int = Field(default=0, ge=0, le=100)

    @property
    def pct_sum_valid(self) -> bool:
        return self.primary_commission_pct + self.secondary_commission_pct == 100


class DealUpdateRequest(BaseModel):
    notes: str | None = None
    expected_close_at: date | None = None
    listing_id: uuid.UUID | None = None
    transaction_value: Decimal | None = None
    transaction_currency: str | None = None
    secondary_agent_id: uuid.UUID | None = None
    primary_commission_pct: int | None = Field(default=None, ge=0, le=100)
    secondary_commission_pct: int | None = Field(default=None, ge=0, le=100)


class DealTransitionRequest(BaseModel):
    stage: DealStage
    description: str | None = None


class DealCommissionOverrideRequest(BaseModel):
    commission_type: CommissionType
    commission_value: Decimal
    override_reason: str = Field(min_length=5)


class DealCommissionPayoutRequest(BaseModel):
    amount: Decimal
    paid_at: datetime | None = None


class DealCommissionPayoutCancelRequest(BaseModel):
    reason_note: str = Field(..., min_length=5, max_length=1000)


class CustomerSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    full_name: str
    phone: str | None = None
    phone_normalized: str | None = None


class AgentSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    full_name: str
    email: str
    default_commission_type: CommissionType | None = None
    default_commission_value: Decimal | None = None


class PropertySummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    title: str
    status: str


class DealResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    tenant_id: uuid.UUID
    deal_type: DealType
    customer_id: uuid.UUID
    property_id: uuid.UUID
    listing_id: uuid.UUID | None = None
    primary_agent_id: uuid.UUID
    secondary_agent_id: uuid.UUID | None = None
    lead_id: uuid.UUID | None = None
    stage: DealStage
    transaction_value: Decimal
    transaction_currency: str
    expected_close_at: date | None = None
    closed_at: datetime | None = None
    notes: str | None = None
    commission_type: CommissionType
    commission_value: Decimal
    commission_amount: Decimal  # computed; not a DB column
    commission_overridden: bool
    commission_override_reason: str | None = None
    commission_payout_status: CommissionPayoutStatus
    commission_paid_amount: Decimal
    commission_paid_at: datetime | None = None
    primary_commission_pct: int = 100
    secondary_commission_pct: int = 0
    admin_confirmed_at: datetime | None = None
    admin_confirmed_by_user_id: uuid.UUID | None = None
    agent_confirmed_at: datetime | None = None
    agent_confirmed_by_user_id: uuid.UUID | None = None
    created_by_user_id: uuid.UUID | None = None
    customer: CustomerSummary | None = None
    primary_agent: AgentSummary | None = None
    interest_property: PropertySummary | None = None
    created_at: datetime
    updated_at: datetime


class DealListResponse(BaseModel):
    items: list[DealResponse]
    total: int

File: apps/web/src/lib/types/deal.ts
import type { CommissionType } from "./agent";

export type DealType = "sale" | "rent_short" | "rent_long";
export type DealStage =
  | "initiated" | "documents_pending" | "deposit_pending"
  | "closed_won" | "closed_lost" | "canceled";
export type CommissionPayoutStatus = "pending" | "partial" | "paid" | "canceled";

export interface DealCustomerSummary {
  id: string;
  full_name: string;
  phone: string | null;
  phone_normalized: string | null;
}

export interface DealAgentSummary {
  id: string;
  full_name: string;
  email: string;
}

export interface DealPropertySummary {
  id: string;
  title: string;
  status: string;
}

export interface Deal {
  id: string;
  tenant_id: string;
  deal_type: DealType;
  customer_id: string;
  property_id: string;
  listing_id: string | null;
  primary_agent_id: string;
  secondary_agent_id: string | null;
  lead_id: string | null;
  stage: DealStage;
  transaction_value: string;
  transaction_currency: string;
  expected_close_at: string | null;
  closed_at: string | null;
  notes: string | null;
  commission_type: CommissionType;
  commission_value: string;
  commission_amount: string;
  commission_overridden: boolean;
  commission_override_reason: string | null;
  commission_payout_status: CommissionPayoutStatus;
  commission_paid_amount: string;
  commission_paid_at: string | null;
  primary_commission_pct: number | null;
  secondary_commission_pct: number | null;
  admin_confirmed_at: string | null;
  admin_confirmed_by_user_id: string | null;
  agent_confirmed_at: string | null;
  created_by_user_id: string;
  customer: DealCustomerSummary | null;
  primary_agent: DealAgentSummary | null;
  secondary_agent: DealAgentSummary | null;
  interest_property: DealPropertySummary | null;
  created_at: string;
  updated_at: string;
}

export interface DealCreateRequest {
  deal_type: DealType;
  customer_id: string;
  property_id: string;
  primary_agent_id: string;
  secondary_agent_id?: string | null;
  primary_commission_pct?: number | null;
  secondary_commission_pct?: number | null;
  transaction_value: string | number;
  transaction_currency: string;
  listing_id?: string | null;
  lead_id?: string | null;
  expected_close_at?: string | null;
  notes?: string | null;
  commission_type?: CommissionType | null;
  commission_value?: string | number | null;
  commission_override_reason?: string | null;
}

export interface DealUpdateRequest {
  notes?: string | null;
  expected_close_at?: string | null;
  listing_id?: string | null;
  transaction_value?: string | number | null;
  transaction_currency?: string | null;
}

export interface DealTransitionRequest {
  stage: DealStage;
  description?: string | null;
}

export interface DealCommissionOverrideRequest {
  commission_type: CommissionType;
  commission_value: string | number;
  override_reason: string;
}

export interface DealCommissionPayoutRequest {
  amount: string | number;
  paid_at?: string | null;
}

export interface DealListResponse {
  items: Deal[];
  total: number;
}

export interface DealListParams {
  stage?: DealStage;
  deal_type?: DealType;
  primary_agent_id?: string;
  customer_id?: string;
  property_id?: string;
  q?: string;
  skip?: number;
  limit?: number;
}

export interface LeaderboardRow {
  agent_id: string;
  agent_name: string;
  deals_won_count: number;
  sold_count: number;
  rented_count: number;
  total_transaction_value: number;
  total_commission_value: number;
  paid_commission: number;
  pending_commission: number;
}

export interface LeaderboardResponse {
  rows: LeaderboardRow[];
  period: string;
}

export type LeaderboardPeriod = "this_month" | "this_quarter" | "this_year" | "custom";

export const DEAL_STAGE_TRANSITIONS: Record<DealStage, DealStage[]> = {
  initiated: ["documents_pending", "canceled", "closed_lost"],
  documents_pending: ["deposit_pending", "initiated", "canceled", "closed_lost"],
  deposit_pending: ["closed_won", "documents_pending", "canceled", "closed_lost"],
  closed_won: [],
  closed_lost: [],
  canceled: [],
};

export const DEAL_TERMINAL_STAGES = new Set<DealStage>(["closed_won", "closed_lost", "canceled"]);
